Fast `series` method for ring series

shivamvats opened the pull request:
This PR implements a `series_fast` method that takes an arbitrary `Basic` expression and expands it using sparse ring manipulations. We tried using `EX` domain for all expansion in #9614, but that proved to be too slow for our liking. This PR starts expanding over the `QQ` domain and keeps adding generators as and when required. Thus, we always work on the simplest (almost) possible domain, resulting in considerable speed-up.

The timings are extremely encouraging:

```
In [31]: %timeit series_fast(sin(a**2) + cos(a)*sin(a),a,100)
1 loops, best of 3: 303 ms per loop

In [32]: %timeit (sin(a**2) + cos(a)*sin(a)).series(a,0,100)
1 loops, best of 3: 5.42 s per loop
```

Because of the way we add generators, there is a constant overhead in expanding a given expression(irrespective of the asked order), depending on the complexity of the expression. So, for small orders, `series_fast` (yes, we need a new name!) might seem slow, but soon far outperforms `series` for even medium sized orders:

```
In [34]: %timeit series_fast(p, a, 10)
1 loops, best of 3: 764 ms per loop

In [35]: %timeit p.series(a,0,10)
1 loops, best of 3: 561 ms per loop
```

Here `series_fast` is actually slower than `series` for 10 terms, but once we expand till the 100th order:

```
In [36]: %timeit series_fast(p, a, 100)
1 loops, best of 3: 917 ms per loop

In [37]: %timeit p.series(a,0,100)
1 loops, best of 3: 6.67 s per loop
```

Clearly, the performance advantage will grow larger with larger orders.

There is a lot of scope for optimising the code (my priority was to make it work). It also needs to be played with to find bugs. However, I feel this is the way to go.

cc @certik @pernici @jksuom @thilinarmtb

ToDo
Fix wrong expansion of 
- [x]  `sin(a**2 - a + 2)**5*cos(a**3 - a/2)`
- [x] `sin(x + a + b)` 
- [ ] Rational exponent in argument: `sin(x**QQ(1,2))`. `_parallel_dict_from_expr_if_gens` etc need to be modified

shivamvats committed:
Allow rs_sin and rs_cos to add required generators to ring

shivamvats committed:
Add series_fast function

shivamvats committed:
Use recursion to expand any expression by expanding each term and combining

shivamvats committed:
Handle function/expressions with constant value

shivamvats committed:
Add compose and add_gens methods to PolyRing and add tests

shivamvats committed:
Add tests and use compose method

certik commented:
Great job! How does this compare to @pernici's fastest code? Per his recent [comment](https://github.com/sympy/sympy/pull/9614#issuecomment-127190514), it looks like this might be faster? I thought @pernici's code was performing well, but that was only for cases where general coefficients (EX) were not needed?

shivamvats committed:
Write _get_constant term function to find constant term

shivamvats committed:
Add rs_tan

shivamvats commented:
I benchmarked some expressions [here](https://gist.github.com/shivamvats/85ece88a20401f19dbb2). The benchmarked expressions are mostly random. Is there a standard set of series for proper benchmarking?

`series_fast` is faster than or as fast as `taylor` in most cases and slower in some. Though I don't fully understand how `taylor` works, my guess is that `taylor` is much slower than `series_fast` when it uses `EX`. As long as `taylor` uses `QQ`, there can't be more than a constant difference in performance as `series_fast` uses `QQ` at all times (i.e, the backend is the same).

In cases in which `series_fast` is slower than `taylor` by more than 100%, it seems to be because `series_fast` had to re-expand as it did not get the required precision the first time. We can use the methods used in `taylor ` to predict the precision functions need to be expanded upto.

certik commented:
Can you point me to which `taylor` you were using?

certik commented:
Also, some tests fail.

shivamvats commented:
[This one](https://github.com/pernici/sympy/tree/rs3)

I had to change `from sympy.polys.monomialtools` to `from sympy.polys.orderings` at line 10 in `rs_taylor.py` to make it work.

certik commented:
I see. The benchmarks are good enough. I am fine with this. Let's finish this up and get it in.

shivamvats committed:
Fix test error

shivamvats committed:
Fix failing tests and disable expand with sring

1. rs_min_pow finds out the index of the required gen
2 Disable expand so that exp(a+b) is not expanded to exp(a)*exp(b)

shivamvats committed:
Remove series_fast_old and rename series_fast to rs_series

shivamvats commented:
@certik @pernici The `series` function now works for taylor series of sin, cos, exp and tan. I have fixed the known bugs. I will gradually add other functions. 

For puiseux series, we need to modify `polys` (as I explained in the email). Other than that this is ready for review.

shivamvats committed:
Bug fix and remove trailing new line

certik commented on sympy/polys/tests/test_ring_series.py:
```diff
@@ -578,3 +580,39 @@ def test_puiseux2():
+def test_rs_series():
+    x, a, b, c = symbols('x, a, b, c')
+
+    assert (rs_series(a, a, 5)).as_expr() == a
```
You do not need a parenthesis around `rs_series`:
```python
rs_series(a, a, 5).as_expr()
```

certik commented on sympy/polys/tests/test_ring_series.py:
```diff
@@ -578,3 +580,39 @@ def test_puiseux2():
+
+    assert (rs_series(a, a, 5)).as_expr() == a
+    assert rs_series(sin(1/a), a, 5).as_expr() == sin(1/a)
+    assert (rs_series(sin(a), a, 5)).as_expr() == (sin(a).series(a, 0,
```
the same here

certik commented on sympy/polys/tests/test_ring_series.py:
```diff
@@ -578,3 +580,39 @@ def test_puiseux2():
+    assert rs_series(sin(1/a), a, 5).as_expr() == sin(1/a)
+    assert (rs_series(sin(a), a, 5)).as_expr() == (sin(a).series(a, 0,
+        5)).removeO()
+    assert (rs_series(sin(a) + cos(a), a, 5)).as_expr() == ((sin(a) +
```
and here

certik commented on sympy/polys/tests/test_ring_series.py:
```diff
@@ -578,3 +580,39 @@ def test_puiseux2():
+        5)).removeO()
+    assert (rs_series(sin(a) + cos(a), a, 5)).as_expr() == ((sin(a) +
+        cos(a)).series(a, 0, 5)).removeO()
+    assert (rs_series(sin(a)*cos(a), a, 5)).as_expr() == ((sin(a)*
```
and here

certik commented on sympy/polys/ring_series.py:
```diff
@@ -1245,8 +1266,13 @@ def rs_tan(p, x, prec):
-                                      "this domain.")
+                R = R.add_gens([tan(c_expr, )])
+                #gens = list(R.gens).remove(x.set_ring(R))
+                #R.drop_to_ground(gens)
```
Why are the two lines commented out?

shivamvats replied:
That's another option I had considered- to move the generators to the coefficient ring. I should probably remove these lines.

certik replied:
Either remove them or document that this is another option to consider and pros/cons.

certik commented on sympy/polys/ring_series.py:
```diff
@@ -1341,8 +1366,13 @@ def rs_sin(p, x, prec):
-                                      "this domain.")
+                R = R.add_gens([sin(c_expr), cos(c_expr)])
+                #gens = list(R.gens).remove(x.set_ring(R))
+                #R.drop_to_ground(gens)
```
Why are the lines commented out?

certik commented on sympy/polys/ring_series.py:
```diff
@@ -1407,8 +1436,12 @@ def rs_cos(p, x, prec):
-                                      "this domain.")
+                R = R.add_gens([sin(c_expr), cos(c_expr)])
+                #gens = list(R.gens).remove(x.set_ring(R))
+                #R.drop_to_ground(gens)
```
why are the lines commented out?

certik commented on sympy/polys/ring_series.py:
```diff
@@ -1419,7 +1452,13 @@ def rs_cos(p, x, prec):
-        return rs_cos(p1, x, prec)*t2 - rs_sin(p1, x, prec)*t1
+        p_cos = rs_cos(p1, x, prec)
+        p_sin = rs_sin(p1, x, prec)
+        R = (R.compose(p_cos.ring)).compose(p_sin.ring)
```
You can remove the parentheses:
```python
 R = R.compose(p_cos.ring).compose(p_sin.ring)
```

certik commented on sympy/polys/ring_series.py:
```diff
@@ -1748,3 +1787,118 @@ def rs_compose_add(p1, p2):
+
+
+def _rs_series(expr, series_rs, a, prec):
+    # XXX Use _parallel_dict_from_expr instead of sring.
```
I would change XXX to TODO, and perhaps explain more about the background why the TODO item is needed, so that other people can pick it up if needed.

certik commented on sympy/polys/ring_series.py:
```diff
@@ -1748,3 +1787,118 @@ def rs_compose_add(p1, p2):
+            raise NotImplementedError
+        R1, series = sring(arg, domain=QQ, expand=False)
+        series_inner = _rs_series(arg, series, a, prec)
+        R = (R.compose(R1)).compose(series_inner.ring)
```
remove parentheses.

certik commented on sympy/polys/ring_series.py:
```diff
@@ -1748,3 +1787,118 @@ def rs_compose_add(p1, p2):
+    else:
+        raise NotImplementedError
+
+def rs_series(expr, a, prec):
```
Since this is a higher level function, it needs a docstring with documentation and a few doctests showing how to use it.

certik commented:
I left some comments. In general this looks good, let's fix the issues I pointed out and merge.

shivamvats committed:
Add docstring and polish code

shivamvats committed:
Add comments and fix doctest

shivamvats commented:
I've made the changes.

shivamvats commented:
@certik Anything else to be added here?

certik commented:
I think it looks good, thanks!

shivamvats commented:
Thanks for the review!
